[AutoFill Debugging] Filter out irrelevant content in text extraction results

whsieh opened the pull request:
#### 49940e975da4aa9d1844cbac77996b810a8c4968
<pre>
[AutoFill Debugging] Filter out irrelevant content in text extraction results
<a href="https://bugs.webkit.org/show_bug.cgi?id=298603">https://bugs.webkit.org/show_bug.cgi?id=298603</a>
<a href="https://rdar.apple.com/160196303">rdar://160196303</a>

Reviewed by Abrar Rahman Protyasha.

Add a way to filter out irrelevant — i.e., non-AutoFill-related — text from text extraction results
using a simple on-device model (text classifier). See below for more details.

* Source/WebKit/Configurations/WebKit.xcconfig:

Weakly link against `CoreML.framework` and `NaturalLanguage.framework`.

* Source/WebKit/Platform/Logging.h:
* Source/WebKit/Platform/classifier/cocoa/TextExtractionFilter.h: Copied from Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.h.
* Source/WebKit/Platform/classifier/cocoa/TextExtractionFilter.mm: Added.
(WebKit::singleton):
(WebKit::TextExtractionFilter::singleton):
(WebKit::TextExtractionFilter::singletonIfCreated):
(WebKit::TextExtractionFilter::TextExtractionFilter):

Add a helper class that wraps the model, exposing a single method to asynchronously classify the
given text. This helper class loads the model and handles evaluation off the main thread (i.e. on
`m_modelQueue`) to avoid impacting UI responsiveness.

(WebKit::TextExtractionFilter::initializeModelIfNeeded):
(WebKit::TextExtractionFilter::prewarm):

Can be used to prewarm the model by kickstarting model initialization — this process involves
compiling the `mlmodel` in bundle resources into an `mlmodelc` if necessary, caching it, and then
loading the compiled model.

(WebKit::TextExtractionFilter::resetCache):

Add a method to reset the prediction cache.

(WebKit::TextExtractionFilter::shouldFilter):

Main entry point into this helper class. This first segments the given text into ~120-length chunks,
and sends each chunk through the model, flagging the entire result if any chunk is flagged. The
resulting string (i.e. the original text, not each chunk) is then saved in `m_cache`, and reset upon
top-level navigation. Because the strings can be arbitrarily large (and can also be costly to hash),
we instead just cache the string by its hash value and length, to keep each entry in the cache a
fixed size.

(WebKit::TextExtractionFilter::segmentText):

Helper method to split a longer paragraph into multiple chunks (based on `chunkSize`) for
evaluation.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestTextExtraction:completionHandler:]):

Prewarm the filtering model before we issue the text extraction request to the web content process,
and use the `filterText` helper below to post-process the results once they arrive.

* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.mm:
(-[_WKTextExtractionConfiguration init]):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.swift:
(WKTextExtractionTextItem.content):
(WKTextExtractionTextItem.selectedRange):

Make the internal `content` and `selectedRange` properties `readwrite`.

* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtractionInternal.h:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.h:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm:
(WebKit::filterTextRecursive):
(WebKit::filterText):

Recursively traverse the given text extraction root item, and use the filter singleton above to set
the text content of any `WKTextExtractionTextItem` to the empty string if it should be filtered out.
Additionally set the `selectedRange` to `{NSNotFound, 0}`, since it no longer makes sense for the
selected range to index into content that&apos;s been filtered out.

* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKWebView+TextExtraction.swift:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::didCommitLoadForMainFrame):
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::didCommitLoadForMainFrame):

See `resetCache()` above.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/299807@main">https://commits.webkit.org/299807@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7c7ccd2695e3ec1c639e6b0a9c92511d2dedeeec

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120195 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39889 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30540 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126546 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72268 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/bc2041e2-f9a4-4983-ac2f-478b5553685c) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40585 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48466 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91275 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60577 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/be0f12ad-1d0f-46e2-ae69-3d01a2af0eb5) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123147 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32411 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107761 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71828 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/db0cbda9-4f7a-4e49-a605-443cab5951f6) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31444 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70174 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101891 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26052 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129449 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47115 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35745 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99896 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47481 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103946 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99738 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25335 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45194 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23218 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43746 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46977 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52683 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46445 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49792 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48129 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented:
> Maybe some `LOG(TextExtraction, ...)` calls in the model initialization and cache population code could help with future debugging.

Good idea! Will add logs, especially around all the errors.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7c7ccd2695e3ec1c639e6b0a9c92511d2dedeeec. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50517

whsieh committed:
[AutoFill Debugging] Filter out irrelevant content in text extraction results
https://bugs.webkit.org/show_bug.cgi?id=298603
rdar://160196303

Reviewed by Abrar Rahman Protyasha.

Add a way to filter out irrelevant — i.e., non-AutoFill-related — text from text extraction results
using a simple on-device model (text classifier). See below for more details.

* Source/WebKit/Configurations/WebKit.xcconfig:

Weakly link against `CoreML.framework` and `NaturalLanguage.framework`.

* Source/WebKit/Platform/Logging.h:
* Source/WebKit/Platform/classifier/cocoa/TextExtractionFilter.h: Copied from Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.h.
* Source/WebKit/Platform/classifier/cocoa/TextExtractionFilter.mm: Added.
(WebKit::singleton):
(WebKit::TextExtractionFilter::singleton):
(WebKit::TextExtractionFilter::singletonIfCreated):
(WebKit::TextExtractionFilter::TextExtractionFilter):

Add a helper class that wraps the model, exposing a single method to asynchronously classify the
given text. This helper class loads the model and handles evaluation off the main thread (i.e. on
`m_modelQueue`) to avoid impacting UI responsiveness.

(WebKit::TextExtractionFilter::initializeModelIfNeeded):
(WebKit::TextExtractionFilter::prewarm):

Can be used to prewarm the model by kickstarting model initialization — this process involves
compiling the `mlmodel` in bundle resources into an `mlmodelc` if necessary, caching it, and then
loading the compiled model.

(WebKit::TextExtractionFilter::resetCache):

Add a method to reset the prediction cache.

(WebKit::TextExtractionFilter::shouldFilter):

Main entry point into this helper class. This first segments the given text into ~120-length chunks,
and sends each chunk through the model, flagging the entire result if any chunk is flagged. The
resulting string (i.e. the original text, not each chunk) is then saved in `m_cache`, and reset upon
top-level navigation. Because the strings can be arbitrarily large (and can also be costly to hash),
we instead just cache the string by its hash value and length, to keep each entry in the cache a
fixed size.

(WebKit::TextExtractionFilter::segmentText):

Helper method to split a longer paragraph into multiple chunks (based on `chunkSize`) for
evaluation.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestTextExtraction:completionHandler:]):

Prewarm the filtering model before we issue the text extraction request to the web content process,
and use the `filterText` helper below to post-process the results once they arrive.

* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.mm:
(-[_WKTextExtractionConfiguration init]):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.swift:
(WKTextExtractionTextItem.content):
(WKTextExtractionTextItem.selectedRange):

Make the internal `content` and `selectedRange` properties `readwrite`.

* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtractionInternal.h:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.h:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm:
(WebKit::filterTextRecursive):
(WebKit::filterText):

Recursively traverse the given text extraction root item, and use the filter singleton above to set
the text content of any `WKTextExtractionTextItem` to the empty string if it should be filtered out.
Additionally set the `selectedRange` to `{NSNotFound, 0}`, since it no longer makes sense for the
selected range to index into content that's been filtered out.

* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKWebView+TextExtraction.swift:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::didCommitLoadForMainFrame):
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::didCommitLoadForMainFrame):

See `resetCache()` above.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/299807@main
